core/emulate.c

```diff
@@ -1085,6 +1199,13 @@ em_status_t EMCALL em_decode_insn(struct em_context_t *ctxt, const uint8_t *insn
             opcode->decode_src2 = group_opcode->decode_src2;
         }
     }
+    if (flags & INSN_BYTEOP) {
+        ctxt->operand_size = 1;
+    }
+    if (flags & INSN_STACK) {
+        if (ctxt->operand_size == 4 && ctxt->mode == EM_MODE_PROT64)
+            ctxt->operand_size = 8;
```

Comment:
Because in `EM_MODE_PROT64` mode, no `INSN_STACK` instruction (i.e. push/pop) can be done with 32-bit operands. The rules that apply to all other instructions default to 32-bit in this case, see:
https://github.com/intel/haxm/blob/1bb4c59/core/emulate.c#L952-L955

Consequently we need to override the operand size in this specific case.

Reply:
I see! For the record, this exception is specified in Intel SDM Vol. 2A, Section 2.2.1.7: _Default 64-Bit Operand Size_. Also, https://wiki.osdev.org/X86-64_Instruction_Encoding#Usage gives a summary of all instructions that default to 64-bit operand size in 64-bit mode.